README.md
# Micro RSVP
Backend for managing RSVP transaction states.

* [Domain](#domain)
  * [States](#states)
* [API](#api)
  * [POST /invitee/new](#post-inviteenew)
  * [GET /invitee/`<code>`](#get-inviteecode)
  * [PUT /invitee/`<code>`/metadata](#put-inviteecodemetadata)
  * [PUT /invitee/`<code>`/rsvp](#put-inviteecodersvp)
  * [PUT /invitee/`<code>`/details](#put-inviteecodedetails)

## Domain
Each user is uniquely identified by the code given to the user.  The must enter the correct 4-digit code to be able to RSVP for the event. Once the user confirms his/her RSVP status, a form must be filled.

### States
The possible states for an invitee are:

#### UNVERIFIED
*Trigger*: when entered into the designated Google Sheet, Zapier calls the endpoint that initializes an invitee instance.

#### VERIFIED
*Trigger*: when user enters the valid code.

#### RESPONDED
*Trigger*: when the user responds to the invitation with `YES`, `NO` or `MAYBE`.

#### FORM_SUBMITTED
*Trigger*: when the user submits the form.


## API

### POST /invitee/new

Insert a new invitee into the db. Requires an auth key. In case an invitee with the specified code exists, it will be _replaced_.
#### Request
```
URL: /invitee/new
Method: POST
Headers:
	Authorization: Bearer 23mlwkejf08ads4j
Content-Type: application/json
Body: {
    name: "Alex Ferguson",
    code: "7643",
    date_given:	"Kadima K",
	given_by: "City of Scottsdale",
	pre_entered_company: "Some Company",
	pre_entered_title: "President",
	relationship: "Friend"
}
```
#### Success Response (200 OK)
```
Content-Type: application/json
Body: {
    status: "SUCCESS",
    data: {
		code: 123,
		rsvp_state: "UNVERIFIED"
	},
}
```

#### Failure Response (500 Internal Server Error)
```
Content-Type: application/json
Body: {
    status: "COULD_NOT_ADD_INVITEE",
}
```

### GET /invitee/`<code>`

Given a code, checks if it is valid, and returns the host's name from the database, and the state of the passcode in the flow. If it has already been verified, then the user will be flagged in the db.

#### Request
```
URL: /invitee/1234
Method: GET
```

#### Success Response (200 OK)
**If first time**:
```
Content-Type: application/json
Body: {
    status: "SUCCESS",
	data: {
		name: "Alex Ferguson",
		code: "7643",
		dateGiven:	"",
		givenBy: "City of Scottsdale",
		company: "Some Company",
		title: "President",
		relationship: "Friend",
		originationSource: "Kadima",
		rsvpState: "VERIFIED"
	}
}
```

**If second time, having completed some parts of the flow**:
```
Content-Type: application/json
Body: {
    status: "SUCCESS",
	data: {
		name: "Alex Ferguson",
		code: "7643",
		dateGiven:	"",
		givenBy: "City of Scottsdale",
		company: "Some Company",
		title: "President",
		relationship: "Friend",
		originationSource: "Kadima",
		confirmation: "Maybe",
		rsvpState: "RESPONDED"
	}
}
```
#### Not Found Response (404 Not Found)
```
Content-Type: application/json
Body: {
    status: "NOT_FOUND"
}
```

### PUT /invitee/`<code>`/metadata
Update the prefilled metadata of the invitation.
#### Request
```
URL: /invitee/1234/metadata
Method: PUT
Body: {
	origination_source: "Kadima",
	given_by: "Alexis K"
}
```

#### Success Response (200 OK)
```
Content-Type: application/json
Body: {
    status: "SUCCESS",
}
```

### PUT /invitee/`<code>`/rsvp
Update the RSVP status of the user.
#### Request
```
URL: /invitee/1234/rsvp
Method: PUT
Body: {
	response: "YES",
	rsvp_state: "RESPONDED"
}
```

#### Success Response (200 OK)
```
Content-Type: application/json
Body: {
    status: "SUCCESS",
}
```

### PUT /invitee/`<code>`/details

Enter and save form details entered by the user.
#### Request
```
URL: /invitee/123/details
Method: PUT
Body: {
	name: "ABC",
	company: "DEF",
	title: "VP",
	address: "1 Jl. 234",
	address_2: "2 Jl. 343",
	city: "BWI",
	state: "MD",
	zip: "20012",
	email: "x@y.com",
	mobile_number: "1233434343",
	work_phone: "123123323",
	website: "www.y.com"
}
```

#### Success Response (200 OK)
```
Content-Type: application/json
Body: {
	status: "SUCCESS",
	rsvp_state: "FORM_SUBMITTED",
}
```

### PUT /invitee/`<code>`/additional-invitees
Enter a JSON array of additional invitees.

#### Request
```
URL: /invitee/123/additional-invitees
Method: PUT
Body: [{"name": "Sh7", "email": "a@b.com"}, {"name": "Cr8", "email": "cr8@a.com"}]
```

#### Success Response (200 OK)
```
Content-Type: application/json
Body: {
	status: "SUCCESS",
}
```


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2017-09-09
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2017-09-09
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/rsvp-backend/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/rsvp-backend/compare/0.1.0...0.1.1


src/rsvp_backend/handlers/invitee.clj
(ns rsvp-backend.handlers.invitee
  (:require [cheshire.core :as json]
            [clojure.walk :as clj-walk]
            [ring.util.request :as req]
            [ring.util.response :as res]
            [rsvp-backend.db.invitee :as db]
            [taoensso.timbre :as log]))

(defn new-invitee
  [request]
  (let [query (:query (:params request))
        keywordized-req-body (clj-walk/keywordize-keys (:body request))
        code (Integer/parseInt (:code keywordized-req-body))
        updated-req-body (assoc keywordized-req-body :code code)
        resp (db/create-invitee updated-req-body)]
    (if (= resp "AWS_ERROR")
      (res/status (res/response {:status "AWS_ERROR"}) 500)
      (res/response {:status "SUCCESS" :data {:code code}}))))

(defn get-invitee
  [request]
  (let [code (:code (:route-params request))
        data (db/get-invitee (Integer/parseInt code))]
    (cond
      (nil? data) (res/status (res/response {:status "NOT_FOUND"}) 404)
      (= data "AWS_ERROR") (res/status (res/response {:status "AWS_ERROR"}) 500)
      true (res/response {:status "SUCCESS" :data (assoc data :rsvp_state "VERIFIED")}))))

(defn update-metadata
  [request]
  (let [code (Integer/parseInt (:code (:route-params request)))
        keywordized-req-body (clj-walk/keywordize-keys (:body request))
        resp (db/update-invitee-metadata code keywordized-req-body)]
    (if (= resp "AWS_ERROR")
      (res/status (res/response {:status "AWS_ERROR"}) 500)
      (res/response {:status "SUCCESS"}))))

(defn update-rsvp
  [request]
  (let [code (Integer/parseInt (:code (:route-params request)))
        user-response (:response (clj-walk/keywordize-keys (:body request)))
        resp (db/update-rsvp-status code user-response)]
    (if (= resp "AWS_ERROR")
      (res/status (res/response {:status "AWS_ERROR"}) 500)
      (res/response {:status "SUCCESS"}))))

(defn update-details
  [request]
  (let [code (Integer/parseInt (:code (:route-params request)))
        keywordized-req-body (clj-walk/keywordize-keys (:body request))
        resp (db/update-invitee-details code keywordized-req-body)]
    (if (= resp "AWS_ERROR")
      (res/status (res/response {:status "AWS_ERROR"}) 500)
      (res/response {:status "SUCCESS" :rsvp_state "FORM_SUBMITTED"}))))

(defn update-additional-invitees
  [request]
  (let [code (Integer/parseInt (:code (:route-params request)))
        json-string (json/generate-string (clj-walk/keywordize-keys (:body request)) {:pretty true})
        resp (db/update-additional-invitees code {:additional_invitees_json json-string})]
    (if (= resp "AWS_ERROR")
      (res/status (res/response {:status "AWS_ERROR"}) 500)
      (res/response {:status "SUCCESS"}))))
